sockeye/rnn.py

```diff
@@ -55,3 +62,153 @@ def get_stacked_rnn(cell_type: str,
+                                                                    axis=1,
+                                                                    name="%sslice" % name)
+        in_gate = mx.sym.Activation(in_gate, act_type="sigmoid",
+                                    name='%si' % name)
+        forget_gate = mx.sym.Activation(forget_gate, act_type="sigmoid",
+                                        name='%sf' % name)
+        in_transform = mx.sym.Activation(in_transform, act_type="tanh",
+                                         name='%sc' % name)
+        out_gate = mx.sym.Activation(out_gate, act_type="sigmoid",
+                                     name='%so' % name)
+        next_c = mx.sym._internal._plus(forget_gate * states[1], in_gate * in_transform,
+                                        name='%sstate' % name)
+        next_h = mx.sym._internal._mul(out_gate,
+                                       mx.sym.Activation(self._cN.normalize(next_c),
+                                                         act_type="tanh"),
+                                       name='%sout' % name)
+        return next_h, [next_h, next_c]
+
+
+class LayerNormPerGateLSTMCell(mx.rnn.LSTMCell):
+    """
+    Long-Short Term Memory (LSTM) network cell with layer normalization per gate.
+    Based on Jimmy Lei Ba et al: Layer Normalization (https://arxiv.org/pdf/1607.06450.pdf)
+
+    :param num_hidden: number of RNN hidden units. Number of units in output symbol.
+    :param prefix: prefix for name of layers (and name of weight if params is None).
+    :param params: RNNParams or None. Container for weight sharing between cells. Created if None.
+    :param forget_bias: bias added to forget gate, default 1.0. Jozefowicz et al. 2015 recommends setting this to 1.0.
+    :param norm_scale: scale/gain for layer normalization.
+    :param norm_shift: shift/bias after layer normalization.
+    """
+
+    def __init__(self,
+                 num_hidden: int,
+                 prefix: str = 'lstm_',
+                 params: Optional[mx.rnn.RNNParams] = None,
+                 forget_bias: float = 1.0,
+                 norm_scale: float = 1.0,
+                 norm_shift: float = 0.0) -> None:
+        super(LayerNormPerGateLSTMCell, self).__init__(num_hidden, prefix, params, forget_bias)
+        self._norm_layers = list()  # type: List[LayerNormalization]
+        for name in ['i', 'f', 'c', 'o', 's']:
+            scale = self.params.get('%s_shift' % name, shape=(num_hidden,),
+                                    init=mx.init.Constant(value=norm_shift))
+            shift = self.params.get('%s_scale' % name, shape=(num_hidden,),
+                                    init=mx.init.Constant(value=norm_scale if name != "f" else forget_bias))
+            self._norm_layers.append(
+                LayerNormalization(num_hidden, prefix="%s%s" % (self._prefix, name), scale=scale, shift=shift))
+
+    @property
+    def state_info(self):
```

Comment:
now that we want back to inherting from LSTMCell you can remove state_info and _gate_names that were added when inherting from BaseCell.